unit_test/main_function_test.py

```diff
@@ -0,0 +1,263 @@
+import sys
+import os
+import subprocess
+
+import pytest
+
+from cibuildwheel.__main__ import main
+from cibuildwheel import windows, linux, macos
+from cibuildwheel.util import BuildSelector
+from cibuildwheel.environment import ParsedEnvironment
+
+@pytest.fixture
+def argtest():
+    class TestArgs(object):
+        def __call__(self, **kwargs): 
+            self.kwargs = dict(kwargs)
+    return TestArgs()
+
+def not_call_mock(*args, **kwargs):
+    raise RuntimeError("This should never be called")
+
+
+def apply_mock_protection(monkeypatch):
+    monkeypatch.setattr(subprocess, "Popen", not_call_mock)
+    monkeypatch.setattr(windows, "urlopen", not_call_mock)
+    monkeypatch.setattr(windows, "build", not_call_mock)
+    monkeypatch.setattr(linux, "build", not_call_mock)
+    monkeypatch.setattr(macos, "build", not_call_mock)
+    monkeypatch.setattr(os.path, "exists", lambda x: True)
+    monkeypatch.setattr(sys, "argv", ["python", "abcabc"])
+
+
+def test_unknown_platform_non_ci(monkeypatch, capsys):
+    monkeypatch.setattr(os, 'environ', {})
+    apply_mock_protection(monkeypatch)
+    with pytest.raises(SystemExit) as exit:
+        main()
+    assert exit.value.code == 2
+    _, err = capsys.readouterr()
+    assert 'cibuildwheel: Unable to detect platform.' in err
+    assert "cibuildwheel should run on your CI server" in err
+
+
+def test_unknown_platform_on_ci(monkeypatch, capsys):
+    monkeypatch.setattr(os, 'environ', {"CI": "true"})
+    apply_mock_protection(monkeypatch)
+    monkeypatch.setattr(sys, "platform", "Something")
+
+    with pytest.raises(SystemExit) as exit:
+        main()
+    _, err = capsys.readouterr()
+    assert exit.value.code == 2
+    assert 'cibuildwheel: Unable to detect platform from "sys.platform"' in err
+
+
+def test_unknown_platform(monkeypatch, capsys):
+    monkeypatch.setattr(os, 'environ', {"CIBW_PLATFORM": "Something"})
+    apply_mock_protection(monkeypatch)
+    with pytest.raises(SystemExit) as exit:
+        main()
+    _, err = capsys.readouterr()
+    assert exit.value.code == 2
+    assert 'cibuildwheel: Unsupported platform: Something' in err
+
+@pytest.mark.parametrize("system", ["macos", "linux", "windows"])
+@pytest.mark.parametrize("choose_platform_method", ["partameter", "environment"])
+def test_platform_chose(system, choose_platform_method, argtest, monkeypatch):
+    apply_mock_protection(monkeypatch)
+    monkeypatch.setattr(globals()[system], "build", argtest)
+    if choose_platform_method == "partameter":
+        monkeypatch.setattr(sys, "argv", sys.argv + ["--platform", system])
+    else:
+        monkeypatch.setattr(os, 'environ', {"CIBW_PLATFORM": system})
+    main()
+    assert argtest.kwargs["project_dir"] == "abcabc"
+
```

Comment:
I use this  `monkeypatch.setattr(globals()[system], "build", argtest)` 

instead of this `build_func = mocker.patch('cibuildwheel.{}.build'.format(system))`

Because it allow to not increase project dependency. Bout if you ok with this it may be better.